crates/recursor/src/recursor.rs

```diff
@@ -0,0 +1,454 @@
+// Copyright 2015-2022 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::{net::SocketAddr, time::Instant};
+
+use async_recursion::async_recursion;
+use futures_util::{future::select_all, FutureExt};
+
+use lru_cache::LruCache;
+use parking_lot::Mutex;
+use tracing::{debug, info, warn};
+
+use trust_dns_proto::{
+    op::Query,
+    rr::{RData, RecordType},
+};
+use trust_dns_resolver::{
+    config::{NameServerConfig, NameServerConfigGroup, Protocol, ResolverOpts},
+    dns_lru::{DnsLru, TtlConfig},
+    error::ResolveError,
+    lookup::Lookup,
+    name_server::NameServerPool,
+    Name, TokioConnectionProvider, TokioHandle,
+};
+
+use crate::{recursor_pool::RecursorPool, Error, ErrorKind};
+
+/// Set of nameservers by the zone name
+type NameServerCache = LruCache<Name, RecursorPool>;
+
+/// A top down recursive resolver which operates off a list of "hints", this is often the root nodes.
+pub struct Recursor {
+    hints: RecursorPool,
+    name_server_cache: Mutex<NameServerCache>,
+    record_cache: DnsLru,
+}
+
+impl Recursor {
+    /// Construct a new recursor using the list of NameServerConfigs for the hint list
+    ///
+    /// # Panics
+    ///
+    /// This will panic if the hints are empty.
+    pub fn new(hints: impl Into<NameServerConfigGroup>) -> Result<Self, ResolveError> {
```

Comment:
`hints` seems like a weird name, is this used in any standard or other implementations? Suggestions: `roots`, `upstream`.